src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs

```diff
@@ -429,34 +421,24 @@ public override long Position
             [System.Security.SecurityCritical]  // auto-generated_required
             get
             {
-                if (_buffer != null)
-                {
-                    throw new NotSupportedException(Strings.NotSupported_UmsSafeBuffer);
-                }
+                if (_buffer != null) throw new NotSupportedException(Strings.NotSupported_UmsSafeBuffer);
+                if (!_isOpen) throw __Error.GetStreamIsClosed();
 
                 // Use a temp to avoid a race
                 long pos = Interlocked.Read(ref _position);
-                if (pos > _capacity)
-                    throw new IndexOutOfRangeException(Strings.IndexOutOfRange_UMSPosition);
+                if (pos > _capacity) throw new IndexOutOfRangeException(Strings.IndexOutOfRange_UMSPosition);
                 byte* ptr = _mem + pos;
-                if (!_isOpen) throw __Error.GetStreamIsClosed();
+
                 return ptr;
             }
             [System.Security.SecurityCritical]  // auto-generated_required
             set
             {
-                if (_buffer != null)
-                    throw new NotSupportedException(Strings.NotSupported_UmsSafeBuffer);
+                if (_buffer != null) throw new NotSupportedException(Strings.NotSupported_UmsSafeBuffer);
                 if (!_isOpen) throw __Error.GetStreamIsClosed();
+                if (value < _mem) throw new IOException(Strings.IO_SeekBeforeBegin);
 
-                // Note: subtracting pointers returns an Int64.  Working around
-                // to avoid hitting compiler warning CS0652 on this line. 
-                if (new IntPtr(value - _mem).ToInt64() > UnmanagedMemStreamMaxLength)
-                    throw new ArgumentOutOfRangeException("offset", Strings.ArgumentOutOfRange_UnmanagedMemStreamLength);
-                if (value < _mem)
-                    throw new IOException(Strings.IO_SeekBeforeBegin);
-
-                Interlocked.Exchange(ref _position, value - _mem);
+                Seek(value - _mem, SeekOrigin.Begin);
```

Comment:
I don't know why there is this difference, but I am glad you noticed it; this is exactly the problem I would like to avoid by having all seek-like logic in one place, including the preconditions for seeking.

As to the differences in error behavior (the place where exception is thrown), I would propose that we let this one go in the name of making seeking consistent, but I will chat with more people about it later today, as I do also worry about compat. I actually would like to do a major overhaul of this class. I really dislike how it's structured. Reasoning about this code, especially error code paths and boundary conditions is very hard. Part of the reason stems from the fact that it's two components mixed into one class (stream over byte* and stream over SafeBuffer), the way we do precondition checks, and the fact that we do unchecked arithmetic. But if I do such changes, it would be close to impossible for me to verify that error code paths stayed compatible.